ibis/backends/dask/udf.py

```diff
@@ -0,0 +1,161 @@
+        groupings = args[0].index
+        return dd.map_partitions(func, *args_objs).groupby(groupings)
+
+    # Define an execution rule for a simple elementwise Series
+    # function
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(dd.Series, nargs))
+    )
+    def execute_udf_node(op, *args, **kwargs):
+        # We have rewritten op.func to be a closure enclosing
+        # the kwargs, and therefore, we do not need to pass
+        # kwargs here. This is true for all udf execution in this
+        # file.
+        # See ibis.udf.vectorized.UserDefinedFunction
+        if isinstance(op._output_type, dt.Struct):
+            # dask does not know how to glue together Tuple[*pd.Series].
+            # We create a wrapper that performs the "gluing" in each partition
+            def wrapper(func, *cols):
+                raw = func(*cols)
+                return pandas.Series(zip(*raw))
+
+            df = dd.map_partitions(wrapper, op.func, *args, meta="object")
+            df.index = args[0].index
+            return df
+        else:
+            df = dd.map_partitions(
+                op.func, *args, meta=op._output_type.to_dask()
+            )
+
+            return df
+
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(object, nargs))
+    )
+    def execute_udf_node_non_dask(op, *args, **kwargs):
+        return op.func(*args)
+
+    return scope
+
+
+@pre_execute.register(ops.AnalyticVectorizedUDF)
+@pre_execute.register(ops.AnalyticVectorizedUDF, ibis.client.Client)
+@pre_execute.register(ops.ReductionVectorizedUDF)
+@pre_execute.register(ops.ReductionVectorizedUDF, ibis.client.Client)
+def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
+    input_type = op.input_type
+    nargs = len(input_type)
+
+    # An execution rule to handle analytic and reduction UDFs over
+    # 1) an ungrouped window,
+    # 2) an ungrouped Aggregate node, or
+    # 3) an ungrouped custom aggregation context
+    @execute_node.register(type(op), *(itertools.repeat(dd.Series, nargs)))
+    def execute_udaf_node_no_groupby(op, *args, aggcontext, **kwargs):
+
+        meta = pandas.Series([], dtype=op._output_type.to_dask())
+
+        @dask.delayed
+        def lazy_agg(*series):
+            return aggcontext.agg(series[0], op.func, *series[1:])
+
+        lazy_result = lazy_agg(*args)
+        # Depending on the type of operation, lazy_result is a Delayed that
+        # could become a dd.Series or a dd.core.Scalar
+        if isinstance(op, ops.AnalyticVectorizedUDF):
+            result = dd.from_delayed(lazy_result, meta=meta)
+        else:
+            # lazy_result is a dd.core.Scalar from an ungrouped reduction
+            # computation is forced so that we can work with the scalar
+            # TODO computation
+            out_dtype = op._output_type.to_dask()
+            if out_dtype == object:
```

Comment:
Perhaps better to check if  op._output_type is struct.

dask object type maybe use to present other types, like, string?

Reply:
Yeah - great point, thanks!